// @stackflow/react/dist/index.mjs
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));

// src/activity/useActivity.ts
import { useContext } from "react";

// src/activity/ActivityProvider.tsx
import { createContext } from "react";

// src/utils/isServer.ts
function isServer() {
  return typeof window === "undefined";
}

// src/utils/isBrowser.ts
function isBrowser() {
  return !isServer();
}

// src/utils/makeRef.ts
function makeRef() {
  const ref = {
    value: null
  };
  function get() {
    return ref.value;
  }
  function set(value) {
    ref.value = value;
  }
  return [get, set];
}

// src/utils/noop.ts
var noop = () => {
};

// src/utils/useMemoDeep.ts
import { useEffect, useRef } from "react";
import compare from "react-fast-compare";
function useMemoDeep(next) {
  const previousRef = useRef(next);
  const previous = previousRef.current;
  const isEqual = compare(previous, next);
  useEffect(() => {
    if (!isEqual) {
      previousRef.current = next;
    }
  });
  return isEqual ? previous : next;
}

// src/activity/ActivityProvider.tsx
import { jsx } from "react/jsx-runtime";
var ActivityContext = createContext(null);
var ActivityProvider = ({
  children,
  value
}) => /* @__PURE__ */ jsx(ActivityContext.Provider, { value: useMemoDeep(value), children });
ActivityProvider.displayName = "ActivityProvider";

// src/activity/useActivity.ts
var useActivity = () => useContext(ActivityContext);

// src/activity/useActivityParams.ts
import { useContext as useContext2 } from "react";
function useActivityParams() {
  return useContext2(ActivityContext).params;
}

// src/stack/useStack.ts
import { useContext as useContext3 } from "react";

// src/stack/StackProvider.tsx
import { createContext as createContext2 } from "react";
import { jsx as jsx2 } from "react/jsx-runtime";
var StackContext = createContext2(null);
var StackProvider = ({
  children,
  value
}) => /* @__PURE__ */ jsx2(StackContext.Provider, { value: useMemoDeep(value), children });
StackProvider.displayName = "StackProvider";

// src/stack/useStack.ts
var useStack = () => useContext3(StackContext);

// src/stackflow.tsx
import { makeCoreStore, makeEvent } from "@stackflow/core";
import { memo, useMemo as useMemo3 } from "react";

// src/activity/makeActivityId.ts
import { id } from "@stackflow/core";
function makeActivityId() {
  return id();
}

// src/activity/makeStepId.ts
import { id as id2 } from "@stackflow/core";
function makeStepId() {
  return id2();
}

// src/core/CoreProvider.tsx
import { createContext as createContext3 } from "react";

// src/shims/useDeferredValue.ts
import React2 from "react";
var _a;
var useDeferredValue = (_a = React2.useDeferredValue) != null ? _a : (value) => value;

// src/shims/useSyncExternalStore.ts
import React3, { useEffect as useEffect2, useState } from "react";
var _a2;
var useSyncExternalStore = (_a2 = React3.useSyncExternalStore) != null ? _a2 : (subscribe, getSnapshot) => {
  const [state, setState] = useState(() => getSnapshot());
  useEffect2(() => {
    const dispose = subscribe(() => {
      setState(() => getSnapshot());
    });
    return dispose;
  });
  return state;
};

// src/shims/useTransition.ts
import React4 from "react";
var _a3;
var useTransition = (_a3 = React4.useTransition) != null ? _a3 : () => [false, (cb) => cb()];

// src/core/CoreProvider.tsx
import { jsx as jsx3 } from "react/jsx-runtime";
var CoreActionsContext = createContext3(
  null
);
var CoreStateContext = createContext3(null);
var CoreProvider = ({
  coreStore,
  children
}) => {
  const stack = useSyncExternalStore(
    coreStore.subscribe,
    coreStore.actions.getStack,
    coreStore.actions.getStack
  );
  const deferredStack = useDeferredValue(stack);
  return /* @__PURE__ */ jsx3(CoreStateContext.Provider, { value: deferredStack, children: /* @__PURE__ */ jsx3(CoreActionsContext.Provider, { value: coreStore.actions, children }) });
};
CoreProvider.displayName = "CoreProvider";

// src/core/useCoreActions.ts
import { useContext as useContext4 } from "react";
var useCoreActions = () => useContext4(CoreActionsContext);

// src/core/useCoreState.ts
import { useContext as useContext5 } from "react";
var useCoreState = () => useContext5(CoreStateContext);

// src/MainRenderer.tsx
import { useEffect as useEffect3 } from "react";

// src/plugins/PluginsProvider.tsx
import { createContext as createContext4 } from "react";
import { jsx as jsx4 } from "react/jsx-runtime";
var PluginsContext = createContext4(null);
var PluginsProvider = ({
  children,
  value
}) => /* @__PURE__ */ jsx4(PluginsContext.Provider, { value, children });
PluginsProvider.displayName = "PluginsProvider";

// src/plugins/usePlugins.ts
import { useContext as useContext6 } from "react";
function usePlugins() {
  return useContext6(PluginsContext);
}

// src/PluginRenderer.tsx
import { jsx as jsx5 } from "react/jsx-runtime";
var PluginRenderer = ({
  activityComponentMap,
  plugin,
  initialContext
}) => {
  const coreState = useCoreState();
  const plugins = usePlugins();
  return plugin.render({
    stack: __spreadProps(__spreadValues({}, coreState), {
      render(overrideStack) {
        const stack = __spreadValues(__spreadValues({}, coreState), overrideStack);
        return {
          activities: stack.activities.map((activity) => __spreadProps(__spreadValues({}, activity), {
            key: activity.id,
            render(overrideActivity) {
              const Activity = activityComponentMap[activity.name];
              let output = /* @__PURE__ */ jsx5(Activity, { params: activity.params });
              plugins.forEach((p) => {
                var _a4, _b;
                output = (_b = (_a4 = p.wrapActivity) == null ? void 0 : _a4.call(p, {
                  activity: __spreadProps(__spreadValues({}, activity), {
                    render: () => output
                  }),
                  initialContext
                })) != null ? _b : output;
              });
              return /* @__PURE__ */ jsx5(StackProvider, { value: stack, children: /* @__PURE__ */ jsx5(
                ActivityProvider,
                {
                  value: __spreadValues(__spreadValues({}, activity), overrideActivity),
                  children: output
                },
                activity.id
              ) });
            }
          }))
        };
      }
    }),
    initialContext
  });
};
PluginRenderer.displayName = "PluginRenderer";
var PluginRenderer_default = PluginRenderer;

// src/MainRenderer.tsx
import { Fragment, jsx as jsx6 } from "react/jsx-runtime";
var MainRenderer = ({
  activityComponentMap,
  initialContext
}) => {
  const coreState = useCoreState();
  const plugins = usePlugins();
  const renderingPlugins = plugins.filter(
    (plugin) => !!plugin.render
  );
  useEffect3(() => {
    if (renderingPlugins.length === 0) {
      console.warn(
        `Stackflow - There is no rendering plugin, so "<Stack />" doesn't render anything. If you want to render some UI, use "@stackflow/plugin-renderer-basic" or add another rendering plugin.`
      );
    }
  }, [renderingPlugins]);
  let output = /* @__PURE__ */ jsx6(Fragment, { children: renderingPlugins.map((plugin) => /* @__PURE__ */ jsx6(
    PluginRenderer_default,
    {
      activityComponentMap,
      plugin,
      initialContext
    },
    plugin.key
  )) });
  plugins.forEach((plugin) => {
    var _a4, _b;
    output = (_b = (_a4 = plugin.wrapStack) == null ? void 0 : _a4.call(plugin, {
      stack: __spreadProps(__spreadValues({}, coreState), {
        render() {
          return output;
        }
      }),
      initialContext
    })) != null ? _b : output;
  });
  return output;
};
MainRenderer.displayName = "MainRenderer";
var MainRenderer_default = MainRenderer;

// src/useActions.ts
import { useMemo } from "react";
function parseActionOptions(options) {
  if (!options) {
    return { skipActiveState: false };
  }
  const isNullableAnimateOption = options.animate === void 0 || options.animate == null;
  if (isNullableAnimateOption) {
    return { skipActiveState: false };
  }
  return { skipActiveState: !options.animate };
}
function useActions() {
  const coreActions = useCoreActions();
  const [pending] = useTransition();
  return useMemo(
    () => ({
      pending,
      push(activityName, activityParams, options) {
        const activityId = makeActivityId();
        coreActions == null ? void 0 : coreActions.push({
          activityId,
          activityName,
          activityParams,
          skipEnterActiveState: parseActionOptions(options).skipActiveState
        });
        return {
          activityId
        };
      },
      replace(activityName, activityParams, options) {
        var _a4;
        const activityId = makeActivityId();
        coreActions == null ? void 0 : coreActions.replace({
          activityId: (_a4 = options == null ? void 0 : options.activityId) != null ? _a4 : makeActivityId(),
          activityName,
          activityParams,
          skipEnterActiveState: parseActionOptions(options).skipActiveState
        });
        return {
          activityId
        };
      },
      pop(count, options) {
        let _count = 1;
        let _options = {};
        if (typeof count === "object") {
          _options = __spreadValues({}, count);
        }
        if (typeof count === "number") {
          _count = count;
        }
        if (options) {
          _options = __spreadValues({}, options);
        }
        for (let i = 0; i < _count; i += 1) {
          coreActions == null ? void 0 : coreActions.pop({
            skipExitActiveState: i === 0 ? parseActionOptions(_options).skipActiveState : true
          });
        }
      }
    }),
    [coreActions == null ? void 0 : coreActions.push, coreActions == null ? void 0 : coreActions.replace, coreActions == null ? void 0 : coreActions.pop, pending]
  );
}

// src/useStepActions.ts
import { useMemo as useMemo2 } from "react";
var useStepActions = () => {
  const coreActions = useCoreActions();
  const [pending] = useTransition();
  return useMemo2(
    () => ({
      pending,
      stepPush(params) {
        const stepId = makeStepId();
        coreActions == null ? void 0 : coreActions.stepPush({
          stepId,
          stepParams: params
        });
      },
      stepReplace(params) {
        const stepId = makeStepId();
        coreActions == null ? void 0 : coreActions.stepReplace({
          stepId,
          stepParams: params
        });
      },
      stepPop() {
        coreActions == null ? void 0 : coreActions.stepPop({});
      }
    }),
    [
      coreActions == null ? void 0 : coreActions.stepPush,
      coreActions == null ? void 0 : coreActions.stepReplace,
      coreActions == null ? void 0 : coreActions.stepPop,
      pending
    ]
  );
};

// src/stackflow.tsx
import { jsx as jsx7 } from "react/jsx-runtime";
function parseActionOptions2(options) {
  if (!options) {
    return { skipActiveState: false };
  }
  const isNullableAnimateOption = options.animate === void 0 || options.animate == null;
  if (isNullableAnimateOption) {
    return { skipActiveState: false };
  }
  return { skipActiveState: !options.animate };
}
function stackflow(options) {
  var _a4;
  const plugins = ((_a4 = options.plugins) != null ? _a4 : []).flat(Infinity).map((p) => p);
  const pluginInstances = plugins.map((plugin) => plugin());
  const activityComponentMap = Object.entries(options.activities).reduce(
    (acc, [key, Activity]) => __spreadProps(__spreadValues({}, acc), {
      [key]: "component" in Activity ? memo(Activity.component) : memo(Activity)
    }),
    {}
  );
  const enoughPastTime = () => new Date().getTime() - options.transitionDuration * 2;
  const staticCoreStore = makeCoreStore({
    initialEvents: [
      makeEvent("Initialized", {
        transitionDuration: options.transitionDuration,
        eventDate: enoughPastTime()
      }),
      ...Object.entries(options.activities).map(
        ([activityName, Activity]) => makeEvent("ActivityRegistered", __spreadValues({
          activityName,
          eventDate: enoughPastTime()
        }, "component" in Activity ? {
          activityParamsSchema: Activity.paramsSchema
        } : null))
      )
    ],
    plugins: []
  });
  const [getCoreStore, setCoreStore] = makeRef();
  const Stack = memo((props) => {
    const coreStore = useMemo3(() => {
      const prevCoreStore = getCoreStore();
      if (isBrowser() && prevCoreStore) {
        return prevCoreStore;
      }
      const initialPushedEventsByOption = options.initialActivity ? [
        makeEvent("Pushed", {
          activityId: makeActivityId(),
          activityName: options.initialActivity(),
          activityParams: {},
          eventDate: enoughPastTime(),
          skipEnterActiveState: false
        })
      ] : [];
      const initialPushedEvents = pluginInstances.reduce(
        (initialEvents, pluginInstance) => {
          var _a5, _b, _c;
          return (_c = (_b = pluginInstance.overrideInitialEvents) == null ? void 0 : _b.call(pluginInstance, {
            initialEvents,
            initialContext: (_a5 = props.initialContext) != null ? _a5 : {}
          })) != null ? _c : initialEvents;
        },
        initialPushedEventsByOption
      );
      const isInitialActivityIgnored = initialPushedEvents.length > 0 && initialPushedEventsByOption.length > 0 && initialPushedEvents !== initialPushedEventsByOption;
      if (isInitialActivityIgnored) {
        console.warn(
          `Stackflow - Some plugin overrides an "initialActivity" option. The "initialActivity" option you set to "${initialPushedEventsByOption[0].activityName}" in the "stackflow" is ignored.`
        );
      }
      if (initialPushedEvents.length === 0) {
        console.warn(
          `Stackflow - There is no initial activity. If you want to set the initial activity, add the \`initialActivity\` option of the \`stackflow()\` function or add a plugin that sets the initial activity. (e.g. \`@stackflow/plugin-history-sync\`)`
        );
      }
      const store = makeCoreStore({
        initialEvents: [
          ...staticCoreStore.pullEvents(),
          ...initialPushedEvents
        ],
        plugins
      });
      if (isBrowser()) {
        store.init();
        setCoreStore(store);
      }
      return store;
    }, []);
    return /* @__PURE__ */ jsx7(PluginsProvider, { value: pluginInstances, children: /* @__PURE__ */ jsx7(CoreProvider, { coreStore, children: /* @__PURE__ */ jsx7(
      MainRenderer_default,
      {
        activityComponentMap,
        initialContext: props.initialContext
      }
    ) }) });
  });
  Stack.displayName = "Stack";
  return {
    activities: options.activities,
    Stack,
    useFlow: useActions,
    useStepFlow: useStepActions,
    addActivity(activity) {
      if (getCoreStore()) {
        console.warn(
          `Stackflow - \`addActivity()\` API cannot be called after a \`<Stack />\` component has been rendered`
        );
        return;
      }
      activityComponentMap[activity.name] = memo(activity.component);
      staticCoreStore.actions.dispatchEvent("ActivityRegistered", {
        activityName: activity.name,
        activityParamsSchema: activity.paramsSchema,
        eventDate: enoughPastTime()
      });
    },
    addPlugin(plugin) {
      if (getCoreStore()) {
        console.warn(
          `Stackflow - \`addPlugin()\` API cannot be called after a \`<Stack />\` component has been rendered`
        );
        return;
      }
      [plugin].flat(Infinity).map((p) => p).forEach((p) => {
        plugins.push(p);
        pluginInstances.push(p());
      });
    },
    actions: {
      getStack() {
        var _a5, _b;
        return (_b = (_a5 = getCoreStore()) == null ? void 0 : _a5.actions.getStack()) != null ? _b : staticCoreStore.actions.getStack();
      },
      dispatchEvent(name, parameters) {
        var _a5;
        return (_a5 = getCoreStore()) == null ? void 0 : _a5.actions.dispatchEvent(name, parameters);
      },
      push(activityName, activityParams, options2) {
        var _a5;
        const activityId = makeActivityId();
        (_a5 = getCoreStore()) == null ? void 0 : _a5.actions.push({
          activityId,
          activityName,
          activityParams,
          skipEnterActiveState: parseActionOptions2(options2).skipActiveState
        });
        return {
          activityId
        };
      },
      replace(activityName, activityParams, options2) {
        var _a5, _b, _c;
        const activityId = (_a5 = options2 == null ? void 0 : options2.activityId) != null ? _a5 : makeActivityId();
        (_c = getCoreStore()) == null ? void 0 : _c.actions.replace({
          activityId: (_b = options2 == null ? void 0 : options2.activityId) != null ? _b : makeActivityId(),
          activityName,
          activityParams,
          skipEnterActiveState: parseActionOptions2(options2).skipActiveState
        });
        return {
          activityId
        };
      },
      pop(count, options2) {
        var _a5;
        let _count = 1;
        let _options = {};
        if (typeof count === "object") {
          _options = __spreadValues({}, count);
        }
        if (typeof count === "number") {
          _count = count;
        }
        if (options2) {
          _options = __spreadValues({}, options2);
        }
        for (let i = 0; i < _count; i += 1) {
          (_a5 = getCoreStore()) == null ? void 0 : _a5.actions.pop({
            skipExitActiveState: i === 0 ? parseActionOptions2(_options).skipActiveState : true
          });
        }
      },
      stepPush(params) {
        var _a5;
        const stepId = makeStepId();
        return (_a5 = getCoreStore()) == null ? void 0 : _a5.actions.stepPush({
          stepId,
          stepParams: params
        });
      },
      stepReplace(params) {
        var _a5;
        const stepId = makeStepId();
        return (_a5 = getCoreStore()) == null ? void 0 : _a5.actions.stepReplace({
          stepId,
          stepParams: params
        });
      },
      stepPop() {
        var _a5;
        return (_a5 = getCoreStore()) == null ? void 0 : _a5.actions.stepPop({});
      }
    }
  };
}

// src/useActiveEffect.ts
import { useEffect as useEffect4 } from "react";
var useActiveEffect = (effect) => {
  const { isActive } = useActivity();
  useEffect4(() => {
    if (isActive) {
      return effect();
    }
    return noop;
  }, [isActive]);
};

// src/useEnterDoneEffect.ts
import { useEffect as useEffect5 } from "react";
var useEnterDoneEffect = (effect, deps = []) => {
  const { isTop, transitionState } = useActivity();
  useEffect5(() => {
    if (isTop && transitionState === "enter-done") {
      return effect();
    }
    return noop;
  }, [isTop, transitionState, ...deps]);
};

// src/useStep.ts
import { useContext as useContext7 } from "react";
function useStep() {
  var _a4;
  const { steps, id: id3 } = useContext7(ActivityContext);
  return (_a4 = steps.filter((step) => step.id !== id3).at(-1)) != null ? _a4 : null;
}
export {
  stackflow,
  useActions,
  useActiveEffect,
  useActivity,
  useActivityParams,
  useEnterDoneEffect,
  useStack,
  useStep,
  useStepActions
};
